iot/end_to_end_example/device/app/src/main/kotlin/com/example/androidthings/coffeecontrol/CoffeeControlActivity.kt

```diff
@@ -0,0 +1,324 @@
+
+            val payload = JsonObject(mapOf("temperature" to currTemp.toInt()))
+                .toJsonString()
+            val event = TelemetryEvent(
+                payload.toByteArray(), null, TelemetryEvent.QOS_AT_LEAST_ONCE)
+
+            client.publishTelemetry(event)
+
+            deviceHandler.postDelayed(this, 2_000) // Delay 2 secs, repost temp
+        }
+    }
+    private val deviceAnimateRunnable = object : Runnable {
+        override fun run() {
+            val colors = IntArray(deviceRainbow.size)
+            animCounter = animCounter + 1
+
+            if (deviceIsSimulated) {
+                // For testing
+                if (currTemp > 40) {
+                    deviceHeaterOn = false
+                }
+                if (currTemp < 10) {
+                    deviceHeaterOn = true
+                }
+            } else {
+                // Configuration messages are used to set heater state in another runnable.
+            }
+
+            if (!deviceIsConnected) {
+                if (animCounter and 1 == 0) {
+                    for (i in colors.indices) {
+                        colors[6 - i] = Color.rgb(0, 0, 0)
+                    }
+                    currTemp = 0f
+                } else {
+                    for (i in colors.indices) {
+                        colors[6 - i] = Color.rgb(0, 255, 0)
+                    }
+                    currTemp = -999f
+                }
+            } else {
+                if (deviceHeaterOn) {
+                    alphaTweak += 255 / 7
+                    currTemp += .03f
+                    for (i in colors.indices) {
+                        val a = alphaTweak + i * (255 / 7)
+                        colors[i] = Color.rgb(a % 255, 0, 0)
+                    }
+                } else {
+                    currTemp -= .03f
+                    for (i in colors.indices) {
+                        val a = alphaTweak + i * (255 / 7)
+                        colors[6 - i] = Color.rgb(0, 0, a % 255)
+                    }
+                }
+            }
+
+            deviceDisplay?.display(currTemp.toDouble())
+            deviceLedstrip?.write(colors)
+
+            // Trigger loop again in future.
+            if (!deviceIsConnected) {
+                if (animCounter < 6) { // Green blink animation
+                    deviceHandler.postDelayed(this, 250)
+                } else {
+                    animCounter = 0
+                    deviceHandler.postDelayed(this, 1000)
+                }
+            } else {
+                deviceHandler.postDelayed(this, FRAME_DELAY_MS.toLong()) // Normal delay
+            }
+        }
+    }
+
+    private fun onConfigurationReceived(bytes: ByteArray) {
+        if (bytes.size == 0) {
+            Log.d(TAG, "Ignoring empty device config event")
+            return
+        }
+
+        try {
+            val config = Klaxon().parse<CoffeeControlConfig>(bytes.toString(UTF_8))
```

Comment:
consider using `String(bytes)`, which should default to UTF-8 in Kotlin.
http://kotlinlang.org/api/latest/jvm/stdlib/kotlin.text/-string.html